### PR TITLE
METRON-1326: Metron deploy with Kerberos fails on Ambari 2.5 during ES service stop

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

Addresses https://issues.apache.org/jira/browse/METRON-1326

I'm running up full dev now to verify ES and Kibana start/stop/restart.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by mmiklavc** [comment]
full dev still runs Ambari 2.4.2.0, so this approach may not work across 2.4 -> 2.5. ES and Kibana services are running, but they show as stopped in the Ambari UI.

image

**Comment by mmiklavc** [comment]
Upon investigation, this looks ok. But there is a permissions issue with the service script. We have some environments that don't have this issue so I'm looking for an appropriate place to correct this.

```
INFO 2018-01-10 19:51:59,540 PythonReflectiveExecutor.py:65 - Reflective command failed with exception:
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/ambari_agent/PythonReflectiveExecutor.py", line 57, in run_file
    imp.load_source('__main__', script)
  File "/var/lib/ambari-agent/cache/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_master.py", line 94, in <module>
    Elasticsearch().execute()
  File "/usr/lib/python2.6/site-packages/resource_management/libraries/script/script.py", line 280, in execute
    method(env)
  File "/var/lib/ambari-agent/cache/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_master.py", line 82, in status
    raise ExecutionFailed(err_msg, rc, out, err)
ExecutionFailed: Execution of 'service elasticsearch status' returned 126
```

```
[elasticsearch@node1 ambari-agent]$ service elasticsearch status
env: /etc/init.d/elasticsearch: Permission denied
```

**Comment by mmiklavc** [comment]
Eureka, I think we got it. Thanks @cestella for the tag team session on digging through the Ambari source.

**Comment by mmiklavc** [comment]
### Testing

1. Spin up full dev
2. Check that ES and Kibana are up as normal
3. Try the following operations - all should succeed without error in Ambari
   - Restart Kibana
   - Restart ES
   - Stop Kibana
   - Stop ES
   - Start Kibana
   - Start ES

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_master.py [hunk 1]
```diff
@@ -56,8 +59,28 @@ def status(self, env):
...
-        status_cmd = "service elasticsearch status"
-        Execute(status_cmd)
+
+        # return codes defined by LSB
+        # http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html
+        cmd = ('service', 'elasticsearch', 'status')
+	
+        rc, out = shell.call(cmd, sudo=True, quiet=False)
+
+        if rc == 3:
```
Comment (anandsubbu) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_master.py]: On my 12-node CentOS 7, when I hit 'Stop Services', Kibana service stop still failed with the following error:

```
stderr:   /var/lib/ambari-agent/data/errors-753.txt

Traceback (most recent call last):
  File "/var/lib/ambari-agent/cache/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py", line 153, in <module>
    Kibana().execute()
  File "/usr/lib/python2.6/site-packages/resource_management/libraries/script/script.py", line 332, in execute
    self.execute_prefix_function(self.command_name, 'after', env)
  File "/usr/lib/python2.6/site-packages/resource_management/libraries/script/script.py", line 350, in execute_prefix_function
    method(env)
  File "/usr/lib/python2.6/site-packages/resource_management/libraries/script/script.py", line 398, in after_stop
    status_method(env)
  File "/var/lib/ambari-agent/cache/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py", line 118, in status
    raise ExecutionFailed(err_msg, rc, out)
resource_management.core.exceptions.ExecutionFailed: Execution of 'service kibana status' returned 2
stdout:   /var/lib/ambari-agent/data/output-753.txt

2018-01-12 10:08:11,166 - Stop Kibana Master
2018-01-12 10:08:11,166 - Execute['service kibana stop'] {}
2018-01-12 10:08:12,251 - Waiting for actual component stop
2018-01-12 10:08:12,251 - Status of the Master
2018-01-12 10:08:12,251 - call[('service', 'kibana', 'status')] {'sudo': True, 'quiet': False}
2018-01-12 10:08:12,285 - call returned (2, 'kibana is not running')

Command failed after 1 tries
```
This check only looks for exit status 3 alone.

I made a local chance to treat any non-zero exit status as `ComponentIsNotRunning` and it worked fine.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_slave.py [hunk 2]
```diff
@@ -56,8 +59,27 @@ def status(self, env):
...
         Logger.info('Check status of Elasticsearch data node')
-        status_cmd = "service elasticsearch status"
-        Execute(status_cmd)
+
+        # return codes defined by LSB
+        # http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html
+        cmd = ('service', 'elasticsearch', 'status')
+        rc, out = shell.call(cmd, sudo=True, quiet=False)
+
+        if rc == 3:
```
Comment (anandsubbu) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/ELASTICSEARCH/5.6.2/package/scripts/elastic_slave.py]: Same comment as Kibana

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py [hunk 3]
```diff
@@ -93,7 +96,26 @@ def status(self, env):
 
         Logger.info("Status of the Master")
 
-        Execute("service kibana status")
+        # return codes defined by LSB
+        # http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html
+	cmd = ('service', 'kibana', 'status')
+	rc, out = shell.call(cmd, sudo=True, quiet=False)
+
+        if rc == 3:
```
Comment (anandsubbu) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py]: Same comment as Kibana

**Comment by mmiklavc** [comment]
@anandsubbu  did this ever work with CentOS 7 from Apache master? I didn't touch that service_check.py file, and furthermore I see there are hardcoded ports for ES.

**Comment by mmiklavc** [comment]
Oh, I see what that is. This appears to be a neglected piece of code. ES has this but Kibana does not, for instance. The metainfo.xml allows you to add "commandScript" items that are listed in the service drop-down menu. But this is distinct from the status check which provides the green checkbox vs red exclamation appearing next to a service name in the Ambari services list.

This is ES
image

This is Kibana
image

I'd like to treat this as distinct from the status check if that seems reasonable to you. I have to admit, I didn't even know the service check piece existed until now.

**Comment by mmiklavc** [comment]
fwiw, this service check also fails in full dev. Git history shows me as creating the file because I moved the entire directory (5.6.2 name change) as part of the ES upgrade, but this code hasn't been touched since September 2016 per `git log --follow`. I'm creating another ticket for it.

https://issues.apache.org/jira/browse/METRON-1400

**Comment by anandsubbu** [comment]
> I'd like to treat this as distinct from the status check if that seems reasonable to you. I have to admit, I didn't even know the service check piece existed until now.

Sure, it makes sense to address this as a separate issue.

**Comment by cestella** [comment]
I'm +1 on this at this point, but suggest we wait until @anandsubbu gives his +1.

**Comment by anandsubbu** [comment]
+1 thank you @mmiklavc.

Spun up a 12-node Centos 7 cluster with the latest fix and validated the following:
* Did 'Stop All Services' -> no issues seen with ES and Kibana services stop any longer.
* Kerberized the cluster -> the process completed successfully and all services came up fine.

The issues reported in the JIRA ticket now stand resolved.

**Pull request closed without merging**